### PR TITLE
Add allow_commission_decrease_at_any_time feature which will allow

### LINKED ISSUE #33843: Commission decrease should be allowed at any time
#### Problem

The commission_updates_only_allowed_in_first_half_of_epoch feature disallows commission decrease in the second half of epoch.  Given that the purpose of this feature was to prevent 'rug pulls' which are accomplished by *increasing commission* at the end of epoch and then *decreasing commission* at the beginning of next epoch, disallowing decreases during the second half of epoch is unnecessary.

Disallowing commission decreases in the second half of epoch prevents an operational technique that I use with my validator: when I have an operator error that results in reduced stake account rewards for the epoch, I reduce commission to 0% to ensure that stake accounts are not disadvantaged by my actions.  Not being allowed to do this in the second half of an epoch is a real problem for me.

#### Proposed Solution

Add a feature which will enable set-commission logic to first check whether the proposed commission change is a decrease, and if so, do not invoke the "only allow commission change in first half of epoch" rule.

### PR DESCRIPTION
**Pull request by bji**
vote account commission to be lowered at any time during the epoch regardless of the commission_updates_only_allowed_in_first_half_of_epoch feature.  Fixes #33843.

### CODE REVIEW

**Comment by AshwinSekar** [comment]
cc @joncinque the author of the original feature.

**Comment by CriesofCarrots** [comment]
Not to delay progress, but we need to get in the habit of starting with a SIMD for feature-gated changes, like this.
Can you please make some version of https://github.com/solana-labs/solana/issues/33843 into a [SIMD](https://github.com/solana-foundation/solana-improvement-documents)? It would be helpful to suss out what parts of the process are significant impediments to smaller changes.

**Comment by bji** [comment]
> Not to delay progress, but we need to get in the habit of starting with a SIMD for feature-gated changes, like this. Can you please make some version of #33843 into a [SIMD](https://github.com/solana-foundation/solana-improvement-documents)? It would be helpful to suss out what parts of the process are significant impediments to smaller changes.

Is a SIMD really needed for all features?  This is a minor feature that really just "fixes" a shortcoming in an existing feature.  Going through the SIMD process sure is painful ...

**Comment by joncinque** [comment]
While I can certainly appreciate the added hurdle of a SIMD, we need to get into the habit of "feature gates require SIMDs" in almost all cases. This should be an easy one to get approved, and I'll be happy to help out!

**Comment by bji** [comment]
SIMD 0080 has been added to propose this feature.

**Comment by bji** [comment]
> While I can certainly appreciate the added hurdle of a SIMD, we need to get into the habit of "feature gates require SIMDs" in almost all cases. This should be an easy one to get approved, and I'll be happy to help out!

The SIMD has been updated to respond to your comments, so it would be great to get that SIMD accepted now.  Thanks.

File: sdk/src/feature_set.rs
Comment (AshwinSekar) [inline sdk/src/feature_set.rs:728]: @joncinque you want to hold the key for this?

File: programs/vote/src/vote_state/mod.rs
Comment (joncinque) [inline programs/vote/src/vote_state/mod.rs:1402]: nit: I know that the rest of the tests use `new_rand` everywhere, but these don't need to be random, and can safely use `Pubkey::new_unique()`
```suggestion
        let node_pubkey = Pubkey::new_unique();
        let withdrawer_pubkey = Pubkey::new_unique();
```

File: programs/vote/src/vote_state/mod.rs
Comment (joncinque) [inline programs/vote/src/vote_state/mod.rs:1478]: micro-nit: I'm surprised clippy allowed this, but it's more idiomatic to do something like
```suggestion
        assert_eq!(
            borrowed_account
                .get_state::<VoteStateVersions>()
                .unwrap()
                .convert_to_current()
                .commission,
                10
        );
```

File: programs/vote/src/vote_state/mod.rs
Comment (joncinque) [inline programs/vote/src/vote_state/mod.rs:1562]: micro-nit
```suggestion
        // Decrease commission in second half of epoch -- allowed because feature_set allows it
```

File: programs/vote/src/vote_state/mod.rs
Comment (joncinque) [inline programs/vote/src/vote_state/mod.rs:881]: nit: rather than take in the `Arc`, can you change these to the ref? ie
```suggestion
    epoch_schedule: &EpochSchedule,
    clock: &Clock,
```
I typically let the compiler tell me what to do to go from `Arc` to a normal ref, but I think you'll need to call into the function with `sysvar_cache.get_epoch_schedule()?.as_ref()`

File: programs/vote/src/vote_state/mod.rs
Comment (joncinque) [inline programs/vote/src/vote_state/mod.rs:899]: This instruction isn't used all that often, so it's not a huge performance impact, but we should note that it's deserializing the vote account twice. It's being done once in `is_commission_increase`, and again in `update_commission`.

You did the right thing to make the change as small as possible while not introducing potential forking. We could have `is_commission_increase` take in the `VoteState` directly and update this function a bit like:

```suggestion
    let mut vote_state: VoteState = if feature_set.is_active(&feature_set::allow_commission_decrease_at_any_time::id()) {
        let mut vote_state: VoteState = vote_account
            .get_state::<VoteStateVersions>()?
            .convert_to_current();

        if is_commission_increase(&vote_state, commission)
            && feature_set
                .is_active(&feature_set::commission_updates_only_allowed_in_first_half_of_epoch::id())
        {
            if !is_commission_update_allowed(clock.slot, epoch_schedule) {
                return Err(VoteError::CommissionUpdateTooLate.into());
            }
        }
        
        vote_state
    } else {
        if feature_set
            .is_active(&feature_set::commission_updates_only_allowed_in_first_half_of_epoch::id())
        {
            if !is_commission_update_allowed(clock.slot, epoch_schedule) {
                return Err(VoteError::CommissionUpdateTooLate.into());
            }
        }
        vote_account
            .get_state::<VoteStateVersions>()?
            .convert_to_current()
    };
```

I know it's a bit more code, but it isn't ridiculously ugly. What do you think? cc @AshwinSekar for another opinion

File: sdk/src/feature_set.rs
Comment (joncinque) [reply sdk/src/feature_set.rs:728]: Yeah no problem, I ground out `decoMktMcnmiq6t3u7g5BfgcQu91nKZr6RvMYf9z1Jb`

**Review by joncinque — COMMENTED** [review]
Looks great for the most part! Most of my comments are small, along with a way to avoid double deserialization in the function now

File: programs/vote/src/vote_state/mod.rs
Comment (AshwinSekar) [reply programs/vote/src/vote_state/mod.rs:899]: Yeah in the vote program we try to deserialize only once at the beginning of each instruction using either:
```rust
vote_account.get_state::<VoteStateVersions>()?.convert_to_current
```
or `verify_and_get_vote_state`

I think you could just let `is_commission_increase` take a vote state and keep it like how you have it:

```rust
    let mut vote_state: VoteState = vote_account.get_state::<VoteStateVersions>()?.convert_to_current();
    let enforce_commission_update_rule =
        if feature_set.is_active(&feature_set::allow_commission_decrease_at_any_time::id()) {
            is_commission_increase(vote_state, commission)?
        } else {
            true
        };

    #[allow(clippy::collapsible_if)]
    if enforce_commission_update_rule
        && feature_set
            .is_active(&feature_set::commission_updates_only_allowed_in_first_half_of_epoch::id())
    {
        if !is_commission_update_allowed(clock.slot, epoch_schedule) {
            return Err(VoteError::CommissionUpdateTooLate.into());
        }
    }
```

Comment (joncinque) [reply programs/vote/src/vote_state/mod.rs:899]: If the vote state deserialization fails though, that could cause forking between old and new versions. old could fail during the comission check, and new could fail during the earlier deserialization, which means that two validators could give different errors for the same instruction.

Does the returned error affect consensus? I always thought that was the case

Comment (AshwinSekar) [reply programs/vote/src/vote_state/mod.rs:899]: as far as I know the returned error doesn't impact bank hash, we only check `.is_ok()`/`is_err()` in places, but I might be wrong:
https://github.com/solana-labs/solana/blob/bba6ea2d6989f8cc59ecc3a7f5ef02cb915792d9/accounts-db/src/accounts.rs#L1399-L1403

I don't mind being safe and making sure we return the same error, doesn't look too ugly.

**Comment by bji** [comment]
Please complete the review and/or merge the PR.  Thanks.

**Review by joncinque — DISMISSED** [review]
Apologies for the lateness, I didn't see any responses so assumed that you were still working on it.

Looks great, thanks for integrating all of our feedback! @AshwinSekar can you also take a look and approve if you have a moment?

**Comment by joncinque** [comment]
I went ahead and fixed the merge conflicts

**Comment by joncinque** [comment]
> looks fine to me @joncinque just to confirm, you're fine that we might be returning a different error in some scenarios?

Unless I'm missing something, I don't think it's possible to return a different error with this implementation.

Also, I checked with the runtime team, and they said the error code doesn't make a difference for consensus, so we're all good from my side.

**Comment by bji** [comment]
> looks fine to me @joncinque just to confirm, you're fine that we might be returning a different error in some scenarios?

Why would that be specifically?  The way I altered the code to match your feedback should mean that the same errors are returned before and after the change ... unless I'm missing something?

**Pull request merged**
